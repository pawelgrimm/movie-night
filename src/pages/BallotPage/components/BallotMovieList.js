import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import MovieListItem from "../../../components/MovieListItem/MovieListItem";
import { ThumbsUp, ThumbsDown } from "react-feather";
import { getMoviesOnBallot } from "../../../services/firebase/api";
import Loader from "../../../components/Loader/Loader";
import { hydrateMovies } from "../../../services/movie/api";
import { fetchMovie } from "../../../services/movie/actions";
import VideoModal from "../../../containers/Modal/components/VideoModal";

const mapStateToProps = ({ movies }) => ({
  savedMovies: movies,
});

const mapDispatchToProps = (dispatch) => ({
  fetchMovie: (id) => dispatch(fetchMovie(id)),
});

export const BallotMovieList = ({ savedMovies, openModal, fetchMovie }) => {
  const [loaded, setLoaded] = useState(false);
  const [movies, setMovies] = useState([]);
  const [movieIds, setMovieIds] = useState([]);
  const { id } = useParams();
  const openVideoModal = (videos) => openModal(<VideoModal videos={videos} />);

  useEffect(() => {
    getMoviesOnBallot(id).then((ref) => {
      const movieIds = ref.val();
      movieIds.forEach((id) => fetchMovie(id));
      setMovieIds(movieIds);
      setMovies(hydrateMovies(savedMovies, movieIds));
      setLoaded(true);
    });
  }, []);

  useEffect(() => {
    if (loaded) {
      setMovies(hydrateMovies(savedMovies, movieIds));
    }
  }, [savedMovies]);

  if (!loaded) {
    return <Loader />;
  }

  return (
    <div className="list-body">
      {movies.map((movie) => {
        const id = movie.info.id;
        return (
          <MovieListItem
            key={id}
            isLoading={!movie.isSaved}
            movie={movie.info}
            collapsedButtons={collapsedButtons(id)}
            expandedButtons={expandedButtons(
              id,
              movie.info.trailers,
              openVideoModal
            )}
          />
        );
      })}
    </div>
  );
};

const collapsedButtons = (movieId) => makeButtons(movieId, true);

const expandedButtons = (movieId, videos, openModal) => [
  videos && videos.length > 0 && (
    <button
      className="button"
      onClick={(event) => {
        event.stopPropagation();
        openModal(videos);
      }}
    >
      Watch Trailer
    </button>
  ),
  ...makeButtons(movieId, false),
];

const makeButtons = (movieId, collapsed = true) => [
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsUp />
  </button>,
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsDown />
  </button>,
];

export default connect(mapStateToProps, mapDispatchToProps)(BallotMovieList);
