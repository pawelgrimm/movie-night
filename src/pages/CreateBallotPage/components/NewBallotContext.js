import React, { useContext } from "react";
import { pushBallot } from "../../../services/server/api";

export const defaultState = {
  newBallotId: null,
  setNewBallotId: () => {},
};

export const useNewBallot = () => {
  return useContext(NewBallotContext);
};

const NewBallotContext = React.createContext(null);

export default NewBallotContext;
