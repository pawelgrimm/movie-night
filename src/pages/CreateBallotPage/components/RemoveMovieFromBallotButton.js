import React, { useCallback } from "react";
import Button from "../../../components/Button/Button";
import { Trash2 as TrashIcon } from "react-feather";
import { useDispatch } from "react-redux";
import { removeMovie } from "../../../services/ballot/actions";

const useRemoveMovie = (id) => {
  const dispatch = useDispatch();
  return useCallback(() => {
    dispatch(removeMovie(id));
  }, [id]);
};

const RemoveMovieFromBallotButton = ({ id, type }) => {
  const removeMovie = useRemoveMovie(id);
  return (
    <Button
      className={type === "icon" && "movie-item__button--small"}
      type="secondary"
      onClick={(event) => {
        event.stopPropagation();
        removeMovie();
      }}
    >
      {type === "icon" ? <TrashIcon /> : "Remove Movie"}
    </Button>
  );
};

export default RemoveMovieFromBallotButton;
