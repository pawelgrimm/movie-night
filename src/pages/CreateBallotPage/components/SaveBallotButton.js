import React from "react";
import { connect } from "react-redux";
import { saveBallot } from "../../../services/ballot/actions";
import Button from "../../../components/Button/Button";
import Loader from "../../../components/Loader/Loader";

const mapStateToProps = ({ ballot }) => ({
  isSaving: ballot.isSaving,
});

const mapDispatchToProps = (dispatch) => ({
  saveBallot: () => dispatch(saveBallot()),
});

export const SaveBallotButton = ({ isSaving, saveBallot }) => {
  return (
    <Button onClick={saveBallot}>
      {isSaving ? <Loader /> : "Save Ballot"}
    </Button>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(SaveBallotButton);
