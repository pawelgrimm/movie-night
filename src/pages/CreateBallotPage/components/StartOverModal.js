import ConfirmModal from "../../../containers/Modal/components/ConfirmModal";
import React from "react";

const StartOverModal = ({ onConfirm }) => {
  return (
    <ConfirmModal
      title="Start Over"
      message="This action is not reversible. Are you sure you want to
        continue?"
      onConfirm={onConfirm}
    />
  );
};

export default StartOverModal;
