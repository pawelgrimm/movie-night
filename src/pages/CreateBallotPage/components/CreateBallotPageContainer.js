import React from "react";
import { useNewBallot } from "./NewBallotContext";
import { Redirect } from "react-router-dom";

const CreateBallotPageContainer = ({ children }) => {
  const { newBallotId } = useNewBallot();
  if (newBallotId) {
    return <Redirect push to={`dashboard/${newBallotId}`} />;
  }

  return <>{children}</>;
};

export default CreateBallotPageContainer;
