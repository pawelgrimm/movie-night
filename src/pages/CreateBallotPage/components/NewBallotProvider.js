import React, { useState } from "react";
import NewBallotContext from "./NewBallotContext";

const NewBallotProvider = ({ children }) => {
  const [newBallotId, setNewBallotId] = useState(null);
  return (
    <NewBallotContext.Provider value={{ newBallotId, setNewBallotId }}>
      {children}
    </NewBallotContext.Provider>
  );
};

export default NewBallotProvider;
