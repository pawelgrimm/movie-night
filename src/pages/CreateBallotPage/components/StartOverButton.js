import React, { useCallback } from "react";
import Button from "../../../components/Button/Button";
import { useDispatch, useSelector } from "react-redux";
import { openModal as openModalAction } from "../../../containers/Modal/service/actions";
import { clearBallot as clearBallotAction } from "../../../services/ballot/actions";
import StartOverModal from "./StartOverModal";

const StartOverButton = () => {
  const movies = useSelector(({ ballot }) => ballot.movies || []);
  const dispatch = useDispatch();
  const clearBallot = useCallback(() => dispatch(clearBallotAction()), [
    dispatch,
  ]);
  const openModal = useCallback(
    () => dispatch(openModalAction(<StartOverModal onConfirm={clearBallot} />)),
    [clearBallot]
  );

  return (
    <Button type="secondary" onClick={openModal} disabled={movies.length < 1}>
      Start Over
    </Button>
  );
};

export default StartOverButton;
