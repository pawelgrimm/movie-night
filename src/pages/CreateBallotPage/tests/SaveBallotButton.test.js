import { expect, jest } from "@jest/globals";
import { shallow } from "enzyme";
import React from "react";
import { SaveBallotButton } from "../components/SaveBallotButton";
import Loader from "../../../components/Loader/Loader";

test("should render button with text", () => {
  const wrapper = shallow(
    <SaveBallotButton onClick={() => {}}>Some Text</SaveBallotButton>
  );
  expect(wrapper).toMatchSnapshot();
});
test("should render button with loader", () => {
  const wrapper = shallow(
    <SaveBallotButton onClick={() => {}} isSaving={true}>
      Some Text
    </SaveBallotButton>
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.contains(<Loader />)).toBe(true);
});
