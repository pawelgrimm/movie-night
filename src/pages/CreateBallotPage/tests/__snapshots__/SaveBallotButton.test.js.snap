// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render button with loader 1`] = `
<Button>
  <Loader />
</Button>
`;

exports[`should render button with text 1`] = `
<Button>
  Save Ballot
</Button>
`;
